[ `Backend` ] : Remove `drf`

baseplate-admin opened the pull request.

baseplate-admin commented:
Hey @AKindWorld,

After some investigation i am rewriting the backend in django-ninja.

This should allow us to reuse most of our codes. The backend will be ready in 2-3 days

AKindWorld commented:
> After some investigation i am rewriting the backend in django-ninja.
>This should allow us to reuse most of our codes. The backend will be ready in 2-3 days

Alrighty sounds good to me. No worries, take your time on it :)

^^

baseplate-admin commented:
Hey @AKindWorld ,

I have got what i wanted. Expect me to finish the backend sooner-ish

```
{"mal_id": 1, "anime_name": "adsf", "anime_name_japanese": "dfsa", "anime_source": null, "anime_aired_from": "2022-08-04T14:15:42Z", "anime_aired_to": "2022-08-04T14:15:44Z", "anime_banner": null, "anime_cover": null, "anime_synopsis": "fdas", "anime_background": "ads", "anime_rating": "4", "updated": "2022-08-04T14:15:53.947Z", "anime_genres": "http://127.0.0.1:8000/api/v1/anime/info/1/genres"}
```

This was never possible in `drf`

baseplate-admin commented:
```python
def get_anime_info(request, filters: AnimeInfoFilters = Query(...)):
    query_dict = filters.dict(exclude_none=True)
    query_object = Q()

    # Django filters anyone?
    anime_name = query_dict.get("anime_name", None)
    if anime_name:
        query_object &= (
            Q(anime_name_japanese__icontains=anime_name)
            | Q(anime_name__icontains=anime_name)
            | Q(anime_name_synonyms__name__icontains=anime_name)
        )

    anime_genres = query_dict.get("anime_genres", None)
    if anime_genres:
        query = Q()
        for position in anime_genres.split(","):
            query |= Q(anime_genres__name__icontains=position)
        query_object &= query

    anime_themes = query_dict.get("anime_themes", None)
    if anime_themes:
        query = Q()
        for position in anime_themes.split(","):
            query |= Q(anime_themes__name__icontains=position)
        query_object &= query

    anime_studios = query_dict.get("anime_studios", None)
    if anime_studios:
        query = Q()
        for position in anime_themes.split(","):
            query |= Q(anime_studios__name__icontains=position)
        query_object &= query

    anime_producer = query_dict.get("anime_producer", None)
    if anime_producer:
        query = Q()
        for position in anime_themes.split(","):
            query |= Q(anime_producer__name__icontains=position)
        query_object &= query

    query = AnimeInfoModel.objects.filter(query_object)
    if not query:
        raise Http404("No %s matches the given query." % query.model._meta.object_name)

    return query


```

one of the ugliest piece of code that i have written in my life

baseplate-admin commented:
Which Produces something like

```SQL
SELECT 
  DISTINCT "anime_animeinfomodel"."id", 
  "anime_animeinfomodel"."mal_id", 
  "anime_animeinfomodel"."anime_name", 
  "anime_animeinfomodel"."anime_name_japanese", 
  "anime_animeinfomodel"."anime_source", 
  "anime_animeinfomodel"."anime_aired_from", 
  "anime_animeinfomodel"."anime_aired_to", 
  "anime_animeinfomodel"."anime_banner", 
  "anime_animeinfomodel"."anime_cover", 
  "anime_animeinfomodel"."anime_synopsis", 
  "anime_animeinfomodel"."anime_background", 
  "anime_animeinfomodel"."anime_rating", 
  "anime_animeinfomodel"."updated" 
FROM 
  "anime_animeinfomodel" 
  LEFT OUTER JOIN "anime_animeinfomodel_anime_name_synonyms" ON (
    "anime_animeinfomodel"."id" = "anime_animeinfomodel_anime_name_synonyms"."animeinfomodel_id"
  ) 
  LEFT OUTER JOIN "anime_animesynonymmodel" ON (
    "anime_animeinfomodel_anime_name_synonyms"."animesynonymmodel_id" = "anime_animesynonymmodel"."id"
  ) 
  INNER JOIN "anime_animeinfomodel_anime_genres" ON (
    "anime_animeinfomodel"."id" = "anime_animeinfomodel_anime_genres"."animeinfomodel_id"
  ) 
  INNER JOIN "anime_animegenremodel" ON (
    "anime_animeinfomodel_anime_genres"."animegenremodel_id" = "anime_animegenremodel"."mal_id"
  ) 
  INNER JOIN "anime_animeinfomodel_anime_themes" ON (
    "anime_animeinfomodel"."id" = "anime_animeinfomodel_anime_themes"."animeinfomodel_id"
  ) 
  INNER JOIN "anime_animethememodel" ON (
    "anime_animeinfomodel_anime_themes"."animethememodel_id" = "anime_animethememodel"."mal_id"
  ) 
  INNER JOIN "anime_animeinfomodel_anime_producers" ON (
    "anime_animeinfomodel"."id" = "anime_animeinfomodel_anime_producers"."animeinfomodel_id"
  ) 
  INNER JOIN "anime_animeproducermodel" ON (
    "anime_animeinfomodel_anime_producers"."animeproducermodel_id" = "anime_animeproducermodel"."mal_id"
  ) 
WHERE 
  (
    (
      "anime_animeinfomodel"."anime_name" LIKE % Gintama % ESCAPE '\' OR "anime_animeinfomodel"."anime_name_japanese" LIKE %Gintama% ESCAPE ' \ ' OR "anime_animesynonymmodel"."name" LIKE %Gintama% ESCAPE ' \ ') AND ("anime_animegenremodel"."name" LIKE %Action% ESCAPE ' \ ' OR "anime_animegenremodel"."name" LIKE %Hero% ESCAPE ' \ ') AND ("anime_animethememodel"."name" LIKE %Shounen% ESCAPE ' \ ' OR "anime_animethememodel"."name" LIKE %Adventure% ESCAPE ' \ ') AND ("anime_animeproducermodel"."name" LIKE %Shounen% ESCAPE ' \ ' OR "anime_animeproducermodel"."name" LIKE %Adventure% ESCAPE ' \ '))

```

baseplate-admin commented:
Got OpenApi tagged

```json
{
  "openapi": "3.0.2",
  "info": {
    "title": "CoreProjectAPI",
    "version": "1.0.0",
    "description": ""
  },
  "paths": {
    "/api/v1/anime/info": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_info_get_anime_info",
        "summary": "Get Anime Info",
        "parameters": [
          {
            "in": "query",
            "name": "anime_name",
            "schema": {
              "title": "Anime Name",
              "type": "string"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "anime_genres",
            "schema": {
              "title": "Anime Genres",
              "type": "string"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "anime_themes",
            "schema": {
              "title": "Anime Themes",
              "type": "string"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "anime_studio",
            "schema": {
              "title": "Anime Studio",
              "type": "string"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "anime_producer",
            "schema": {
              "title": "Anime Producer",
              "type": "string"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "limit",
            "schema": {
              "title": "Limit",
              "default": 100,
              "minimum": 1,
              "type": "integer"
            },
            "required": false
          },
          {
            "in": "query",
            "name": "offset",
            "schema": {
              "title": "Offset",
              "default": 0,
              "minimum": 0,
              "type": "integer"
            },
            "required": false
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/PagedAnimeInfoSchema"
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_info_get_individual_anime_info",
        "summary": "Get Individual Anime Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/AnimeInfoSchema"
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}/genres": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_genre_get_individual_anime_genre_info",
        "summary": "Get Individual Anime Genre Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/AnimeGenreSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      },
      "post": {
        "operationId": "apps_api_v1_anime_api_anime_genre_post_individual_anime_genre_info",
        "summary": "Post Individual Anime Genre Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/AnimeGenreSchema"
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/AnimeGenreSchema"
              }
            }
          },
          "required": true
        }
      }
    },
    "/api/v1/anime/info/{anime_id}/producers": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_producer_get_individual_anime_producer_info",
        "summary": "Get Individual Anime Producer Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/AnimeProducerSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}/studios": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_studio_get_individual_anime_studio_info",
        "summary": "Get Individual Anime Studio Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/AnimeStudioSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}/character": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_character_get_individual_anime_character_info",
        "summary": "Get Individual Anime Character Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/AnimeCharacterSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}/themes": {
      "get": {
        "operationId": "apps_api_v1_anime_api_anime_theme_get_individual_anime_theme_info",
        "summary": "Get Individual Anime Theme Info",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/AnimeThemeSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_info"
        ]
      }
    },
    "/api/v1/anime/info/{anime_id}/episodes": {
      "get": {
        "operationId": "apps_api_v1_anime_api_episode_get_individual_anime_episodes",
        "summary": "Get Individual Anime Episodes",
        "parameters": [
          {
            "in": "path",
            "name": "anime_id",
            "schema": {
              "title": "Anime Id",
              "type": "integer"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "title": "Response",
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/EpisodeSchema"
                  }
                }
              }
            }
          }
        },
        "tags": [
          "anime_episodes"
        ]
      }
    },
    "/api/v1/user/info": {
      "get": {
        "operationId": "apps_api_v1_user_api_user_info_get_user_info",
        "summary": "Get User Info",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserSchema"
                }
              }
            }
          }
        },
        "tags": [
          "user"
        ]
      }
    },
    "/api/v1/trackers/mal": {
      "get": {
        "operationId": "apps_api_v1_trackers_api_mal_get_mal_info",
        "summary": "Get Mal Info",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/MALGETSchema"
                }
              }
            }
          }
        },
        "tags": [
          "trackers"
        ]
      },
      "post": {
        "operationId": "apps_api_v1_trackers_api_mal_post_mal_info",
        "summary": "Post Mal Info",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/MALGETSchema"
                }
              }
            }
          }
        },
        "tags": [
          "trackers"
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/MALPOSTSchema"
              }
            }
          },
          "required": true
        }
      }
    },
    "/api/v1/trackers/kitsu": {
      "get": {
        "operationId": "apps_api_v1_trackers_api_kitsu_get_kitsu_info",
        "summary": "Get Kitsu Info",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/KitsuGETSchema"
                }
              }
            }
          }
        },
        "tags": [
          "trackers"
        ]
      },
      "post": {
        "operationId": "apps_api_v1_trackers_api_kitsu_post_kitsu_info",
        "summary": "Post Kitsu Info",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/KitsuGETSchema"
                }
              }
            }
          }
        },
        "tags": [
          "trackers"
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/KitsuPOSTSchema"
              }
            }
          },
          "required": true
        }
      }
    }
  },
  "components": {
    "schemas": {
      "AnimeSynonymSchema": {
        "title": "AnimeSynonymSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "maxLength": 100,
            "type": "string"
          }
        },
        "required": [
          "name"
        ]
      },
      "AnimeInfoSchema": {
        "title": "AnimeInfoSchema",
        "type": "object",
        "properties": {
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "anime_name": {
            "title": "Anime Name",
            "maxLength": 1024,
            "type": "string"
          },
          "anime_name_japanese": {
            "title": "Anime Name Japanese",
            "maxLength": 1024,
            "type": "string"
          },
          "anime_source": {
            "title": "Anime Source",
            "maxLength": 128,
            "type": "string"
          },
          "anime_aired_from": {
            "title": "Anime Aired From",
            "type": "string",
            "format": "date-time"
          },
          "anime_aired_to": {
            "title": "Anime Aired To",
            "type": "string",
            "format": "date-time"
          },
          "anime_banner": {
            "title": "Anime Banner",
            "type": "string"
          },
          "anime_cover": {
            "title": "Anime Cover",
            "type": "string"
          },
          "anime_synopsis": {
            "title": "Anime Synopsis",
            "type": "string"
          },
          "anime_background": {
            "title": "Anime Background",
            "type": "string"
          },
          "anime_rating": {
            "title": "Anime Rating",
            "maxLength": 50,
            "type": "string"
          },
          "updated": {
            "title": "Updated",
            "type": "string",
            "format": "date-time"
          },
          "anime_name_synonyms": {
            "title": "Anime Name Synonyms",
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/AnimeSynonymSchema"
            }
          },
          "anime_genres": {
            "title": "Anime Genres",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          },
          "anime_producers": {
            "title": "Anime Producers",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          },
          "anime_studios": {
            "title": "Anime Studios",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          },
          "anime_characters": {
            "title": "Anime Characters",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          },
          "anime_theme": {
            "title": "Anime Theme",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          },
          "episode": {
            "title": "Episode",
            "minLength": 1,
            "maxLength": 65536,
            "format": "uri",
            "type": "string"
          }
        },
        "required": [
          "mal_id",
          "anime_name",
          "updated",
          "anime_genres",
          "anime_producers",
          "anime_studios",
          "anime_characters",
          "anime_theme",
          "episode"
        ]
      },
      "PagedAnimeInfoSchema": {
        "title": "PagedAnimeInfoSchema",
        "type": "object",
        "properties": {
          "items": {
            "title": "Items",
            "default": [],
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/AnimeInfoSchema"
            }
          },
          "count": {
            "title": "Count",
            "type": "integer"
          }
        },
        "required": [
          "count"
        ]
      },
      "AnimeGenreSchema": {
        "title": "AnimeGenreSchema",
        "type": "object",
        "properties": {
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "default": "",
            "maxLength": 50,
            "type": "string"
          },
          "type": {
            "title": "Type",
            "default": "",
            "maxLength": 50,
            "type": "string"
          }
        }
      },
      "AnimeProducerSchema": {
        "title": "AnimeProducerSchema",
        "type": "object",
        "properties": {
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "default": "",
            "maxLength": 50,
            "type": "string"
          },
          "type": {
            "title": "Type",
            "default": "",
            "maxLength": 50,
            "type": "string"
          }
        }
      },
      "AnimeStudioSchema": {
        "title": "AnimeStudioSchema",
        "type": "object",
        "properties": {
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "default": "",
            "maxLength": 50,
            "type": "string"
          },
          "type": {
            "title": "Type",
            "default": "",
            "maxLength": 50,
            "type": "string"
          }
        }
      },
      "AnimeCharacterSchema": {
        "title": "AnimeCharacterSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "maxLength": 1024,
            "type": "string"
          },
          "character_image": {
            "title": "Character Image",
            "type": "string"
          }
        },
        "required": [
          "mal_id",
          "name"
        ]
      },
      "AnimeThemeSchema": {
        "title": "AnimeThemeSchema",
        "type": "object",
        "properties": {
          "mal_id": {
            "title": "Mal Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "default": "",
            "maxLength": 50,
            "type": "string"
          },
          "type": {
            "title": "Type",
            "default": "",
            "maxLength": 50,
            "type": "string"
          }
        }
      },
      "EpisodeSchema": {
        "title": "EpisodeSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "episode_number": {
            "title": "Episode Number",
            "default": 0,
            "type": "integer"
          },
          "episode_name": {
            "title": "Episode Name",
            "maxLength": 1024,
            "type": "string"
          },
          "episode_cover": {
            "title": "Episode Cover",
            "type": "string"
          },
          "episode_file": {
            "title": "Episode File",
            "type": "string"
          },
          "episode_summary": {
            "title": "Episode Summary",
            "default": "",
            "type": "string"
          },
          "episode_comments": {
            "title": "Episode Comments",
            "type": "array",
            "items": {
              "type": "integer"
            }
          },
          "episode_timestamps": {
            "title": "Episode Timestamps",
            "type": "array",
            "items": {
              "type": "integer"
            }
          }
        },
        "required": [
          "episode_name",
          "episode_comments",
          "episode_timestamps"
        ]
      },
      "UserSchema": {
        "title": "UserSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "last_login": {
            "title": "Last Login",
            "type": "string",
            "format": "date-time"
          },
          "is_superuser": {
            "title": "Superuser Status",
            "description": "Designates that this user has all permissions without explicitly assigning them.",
            "default": false,
            "type": "boolean"
          },
          "username": {
            "title": "Username",
            "description": "Required. 150 characters or fewer. Letters, digits and @/./+/-/_ only.",
            "maxLength": 150,
            "type": "string"
          },
          "first_name": {
            "title": "First Name",
            "maxLength": 150,
            "type": "string"
          },
          "last_name": {
            "title": "Last Name",
            "maxLength": 150,
            "type": "string"
          },
          "email": {
            "title": "Email Address",
            "maxLength": 254,
            "type": "string"
          },
          "is_staff": {
            "title": "Staff Status",
            "description": "Designates whether the user can log into this admin site.",
            "default": false,
            "type": "boolean"
          },
          "is_active": {
            "title": "Active",
            "description": "Designates whether this user should be treated as active. Unselect this instead of deleting accounts.",
            "default": true,
            "type": "boolean"
          },
          "date_joined": {
            "title": "Date Joined",
            "type": "string",
            "format": "date-time"
          },
          "avatar": {
            "title": "Avatar",
            "type": "string"
          },
          "video_volume": {
            "title": "Video Volume",
            "default": 100,
            "type": "integer"
          },
          "groups": {
            "title": "Groups",
            "description": "The groups this user belongs to. A user will get all permissions granted to each of their groups.",
            "type": "array",
            "items": {
              "type": "integer"
            }
          },
          "user_permissions": {
            "title": "User Permissions",
            "description": "Specific permissions for this user.",
            "type": "array",
            "items": {
              "type": "integer"
            }
          }
        },
        "required": [
          "username",
          "groups",
          "user_permissions"
        ]
      },
      "MALGETSchema": {
        "title": "MALGETSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "user": {
            "title": "User",
            "type": "integer"
          },
          "access_token": {
            "title": "Access Token",
            "maxLength": 1024,
            "type": "string"
          },
          "expires_in": {
            "title": "Expires In",
            "type": "number",
            "format": "time-delta"
          },
          "refresh_token": {
            "title": "Refresh Token",
            "maxLength": 1024,
            "type": "string"
          },
          "created_at": {
            "title": "Created At",
            "type": "string",
            "format": "date-time"
          }
        },
        "required": [
          "user"
        ]
      },
      "MALPOSTSchema": {
        "title": "MALPOSTSchema",
        "type": "object",
        "properties": {
          "access_token": {
            "title": "Access Token",
            "maxLength": 1024,
            "type": "string"
          },
          "expires_in": {
            "title": "Expires In",
            "type": "number",
            "format": "time-delta"
          },
          "refresh_token": {
            "title": "Refresh Token",
            "maxLength": 1024,
            "type": "string"
          }
        }
      },
      "KitsuGETSchema": {
        "title": "KitsuGETSchema",
        "type": "object",
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "user": {
            "title": "User",
            "type": "integer"
          },
          "access_token": {
            "title": "Access Token",
            "maxLength": 64,
            "type": "string"
          },
          "expires_in": {
            "title": "Expires In",
            "type": "number",
            "format": "time-delta"
          },
          "refresh_token": {
            "title": "Refresh Token",
            "maxLength": 64,
            "type": "string"
          },
          "created_at": {
            "title": "Created At",
            "type": "string",
            "format": "date-time"
          }
        },
        "required": [
          "user"
        ]
      },
      "KitsuPOSTSchema": {
        "title": "KitsuPOSTSchema",
        "type": "object",
        "properties": {
          "access_token": {
            "title": "Access Token",
            "maxLength": 64,
            "type": "string"
          },
          "expires_in": {
            "title": "Expires In",
            "type": "number",
            "format": "time-delta"
          },
          "refresh_token": {
            "title": "Refresh Token",
            "maxLength": 64,
            "type": "string"
          }
        }
      }
    }
  }
}
```

baseplate-admin commented:
Which gives us this nice typescript file

```typescript
/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  "/api/v1/anime/info": {
    get: operations["apps_api_v1_anime_api_anime_info_get_anime_info"];
  };
  "/api/v1/anime/info/{anime_id}": {
    get: operations["apps_api_v1_anime_api_anime_info_get_individual_anime_info"];
  };
  "/api/v1/anime/info/{anime_id}/genres": {
    get: operations["apps_api_v1_anime_api_anime_genre_get_individual_anime_genre_info"];
    post: operations["apps_api_v1_anime_api_anime_genre_post_individual_anime_genre_info"];
  };
  "/api/v1/anime/info/{anime_id}/producers": {
    get: operations["apps_api_v1_anime_api_anime_producer_get_individual_anime_producer_info"];
  };
  "/api/v1/anime/info/{anime_id}/studios": {
    get: operations["apps_api_v1_anime_api_anime_studio_get_individual_anime_studio_info"];
  };
  "/api/v1/anime/info/{anime_id}/character": {
    get: operations["apps_api_v1_anime_api_anime_character_get_individual_anime_character_info"];
  };
  "/api/v1/anime/info/{anime_id}/themes": {
    get: operations["apps_api_v1_anime_api_anime_theme_get_individual_anime_theme_info"];
  };
  "/api/v1/anime/info/{anime_id}/episodes": {
    get: operations["apps_api_v1_anime_api_episode_get_individual_anime_episodes"];
  };
  "/api/v1/user/info": {
    get: operations["apps_api_v1_user_api_user_info_get_user_info"];
  };
  "/api/v1/trackers/mal": {
    get: operations["apps_api_v1_trackers_api_mal_get_mal_info"];
    post: operations["apps_api_v1_trackers_api_mal_post_mal_info"];
  };
  "/api/v1/trackers/kitsu": {
    get: operations["apps_api_v1_trackers_api_kitsu_get_kitsu_info"];
    post: operations["apps_api_v1_trackers_api_kitsu_post_kitsu_info"];
  };
}

export interface components {
  schemas: {
    /** AnimeSynonymSchema */
    AnimeSynonymSchema: {
      /** Id */
      id?: number;
      /** Name */
      name: string;
    };
    /** AnimeInfoSchema */
    AnimeInfoSchema: {
      /** Mal Id */
      mal_id: number;
      /** Anime Name */
      anime_name: string;
      /** Anime Name Japanese */
      anime_name_japanese?: string;
      /** Anime Source */
      anime_source?: string;
      /**
       * Anime Aired From
       * Format: date-time
       */
      anime_aired_from?: string;
      /**
       * Anime Aired To
       * Format: date-time
       */
      anime_aired_to?: string;
      /** Anime Banner */
      anime_banner?: string;
      /** Anime Cover */
      anime_cover?: string;
      /** Anime Synopsis */
      anime_synopsis?: string;
      /** Anime Background */
      anime_background?: string;
      /** Anime Rating */
      anime_rating?: string;
      /**
       * Updated
       * Format: date-time
       */
      updated: string;
      /** Anime Name Synonyms */
      anime_name_synonyms?: components["schemas"]["AnimeSynonymSchema"][];
      /**
       * Anime Genres
       * Format: uri
       */
      anime_genres: string;
      /**
       * Anime Producers
       * Format: uri
       */
      anime_producers: string;
      /**
       * Anime Studios
       * Format: uri
       */
      anime_studios: string;
      /**
       * Anime Characters
       * Format: uri
       */
      anime_characters: string;
      /**
       * Anime Theme
       * Format: uri
       */
      anime_theme: string;
      /**
       * Episode
       * Format: uri
       */
      episode: string;
    };
    /** PagedAnimeInfoSchema */
    PagedAnimeInfoSchema: {
      /**
       * Items
       * @default []
       */
      items?: components["schemas"]["AnimeInfoSchema"][];
      /** Count */
      count: number;
    };
    /** AnimeGenreSchema */
    AnimeGenreSchema: {
      /** Mal Id */
      mal_id?: number;
      /**
       * Name
       * @default
       */
      name?: string;
      /**
       * Type
       * @default
       */
      type?: string;
    };
    /** AnimeProducerSchema */
    AnimeProducerSchema: {
      /** Mal Id */
      mal_id?: number;
      /**
       * Name
       * @default
       */
      name?: string;
      /**
       * Type
       * @default
       */
      type?: string;
    };
    /** AnimeStudioSchema */
    AnimeStudioSchema: {
      /** Mal Id */
      mal_id?: number;
      /**
       * Name
       * @default
       */
      name?: string;
      /**
       * Type
       * @default
       */
      type?: string;
    };
    /** AnimeCharacterSchema */
    AnimeCharacterSchema: {
      /** Id */
      id?: number;
      /** Mal Id */
      mal_id: number;
      /** Name */
      name: string;
      /** Character Image */
      character_image?: string;
    };
    /** AnimeThemeSchema */
    AnimeThemeSchema: {
      /** Mal Id */
      mal_id?: number;
      /**
       * Name
       * @default
       */
      name?: string;
      /**
       * Type
       * @default
       */
      type?: string;
    };
    /** EpisodeSchema */
    EpisodeSchema: {
      /** Id */
      id?: number;
      /**
       * Episode Number
       * @default 0
       */
      episode_number?: number;
      /** Episode Name */
      episode_name: string;
      /** Episode Cover */
      episode_cover?: string;
      /** Episode File */
      episode_file?: string;
      /**
       * Episode Summary
       * @default
       */
      episode_summary?: string;
      /** Episode Comments */
      episode_comments: number[];
      /** Episode Timestamps */
      episode_timestamps: number[];
    };
    /** UserSchema */
    UserSchema: {
      /** Id */
      id?: number;
      /**
       * Last Login
       * Format: date-time
       */
      last_login?: string;
      /**
       * Superuser Status
       * @description Designates that this user has all permissions without explicitly assigning them.
       * @default false
       */
      is_superuser?: boolean;
      /**
       * Username
       * @description Required. 150 characters or fewer. Letters, digits and @/./+/-/_ only.
       */
      username: string;
      /** First Name */
      first_name?: string;
      /** Last Name */
      last_name?: string;
      /** Email Address */
      email?: string;
      /**
       * Staff Status
       * @description Designates whether the user can log into this admin site.
       * @default false
       */
      is_staff?: boolean;
      /**
       * Active
       * @description Designates whether this user should be treated as active. Unselect this instead of deleting accounts.
       * @default true
       */
      is_active?: boolean;
      /**
       * Date Joined
       * Format: date-time
       */
      date_joined?: string;
      /** Avatar */
      avatar?: string;
      /**
       * Video Volume
       * @default 100
       */
      video_volume?: number;
      /**
       * Groups
       * @description The groups this user belongs to. A user will get all permissions granted to each of their groups.
       */
      groups: number[];
      /**
       * User Permissions
       * @description Specific permissions for this user.
       */
      user_permissions: number[];
    };
    /** MALGETSchema */
    MALGETSchema: {
      /** Id */
      id?: number;
      /** User */
      user: number;
      /** Access Token */
      access_token?: string;
      /**
       * Expires In
       * Format: time-delta
       */
      expires_in?: number;
      /** Refresh Token */
      refresh_token?: string;
      /**
       * Created At
       * Format: date-time
       */
      created_at?: string;
    };
    /** MALPOSTSchema */
    MALPOSTSchema: {
      /** Access Token */
      access_token?: string;
      /**
       * Expires In
       * Format: time-delta
       */
      expires_in?: number;
      /** Refresh Token */
      refresh_token?: string;
    };
    /** KitsuGETSchema */
    KitsuGETSchema: {
      /** Id */
      id?: number;
      /** User */
      user: number;
      /** Access Token */
      access_token?: string;
      /**
       * Expires In
       * Format: time-delta
       */
      expires_in?: number;
      /** Refresh Token */
      refresh_token?: string;
      /**
       * Created At
       * Format: date-time
       */
      created_at?: string;
    };
    /** KitsuPOSTSchema */
    KitsuPOSTSchema: {
      /** Access Token */
      access_token?: string;
      /**
       * Expires In
       * Format: time-delta
       */
      expires_in?: number;
      /** Refresh Token */
      refresh_token?: string;
    };
  };
}

export interface operations {
  apps_api_v1_anime_api_anime_info_get_anime_info: {
    parameters: {
      query: {
        anime_name?: string;
        anime_genres?: string;
        anime_themes?: string;
        anime_studio?: string;
        anime_producer?: string;
        limit?: number;
        offset?: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["PagedAnimeInfoSchema"];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_info_get_individual_anime_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeInfoSchema"];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_genre_get_individual_anime_genre_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeGenreSchema"][];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_genre_post_individual_anime_genre_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeGenreSchema"];
        };
      };
    };
    requestBody: {
      content: {
        "application/json": components["schemas"]["AnimeGenreSchema"];
      };
    };
  };
  apps_api_v1_anime_api_anime_producer_get_individual_anime_producer_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeProducerSchema"][];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_studio_get_individual_anime_studio_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeStudioSchema"][];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_character_get_individual_anime_character_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeCharacterSchema"][];
        };
      };
    };
  };
  apps_api_v1_anime_api_anime_theme_get_individual_anime_theme_info: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["AnimeThemeSchema"][];
        };
      };
    };
  };
  apps_api_v1_anime_api_episode_get_individual_anime_episodes: {
    parameters: {
      path: {
        anime_id: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["EpisodeSchema"][];
        };
      };
    };
  };
  apps_api_v1_user_api_user_info_get_user_info: {
    parameters: {};
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["UserSchema"];
        };
      };
    };
  };
  apps_api_v1_trackers_api_mal_get_mal_info: {
    parameters: {};
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["MALGETSchema"];
        };
      };
    };
  };
  apps_api_v1_trackers_api_mal_post_mal_info: {
    parameters: {};
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["MALGETSchema"];
        };
      };
    };
    requestBody: {
      content: {
        "application/json": components["schemas"]["MALPOSTSchema"];
      };
    };
  };
  apps_api_v1_trackers_api_kitsu_get_kitsu_info: {
    parameters: {};
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["KitsuGETSchema"];
        };
      };
    };
  };
  apps_api_v1_trackers_api_kitsu_post_kitsu_info: {
    parameters: {};
    responses: {
      /** OK */
      200: {
        content: {
          "application/json": components["schemas"]["KitsuGETSchema"];
        };
      };
    };
    requestBody: {
      content: {
        "application/json": components["schemas"]["KitsuPOSTSchema"];
      };
    };
  };
}

export interface external {}

```

baseplate-admin commented:
@AKindWorld  do we implement the search user feature?

baseplate-admin commented:
Hey @AKindWorld,

The backend is finished :) 

Please merge this

AKindWorld commented:
Hi! sorry for the extremely delayed reply.

Well done on the backend part, the code looks amazing (yet I can't comprehend it at all), good job!

Also, by user-search, do you mean being able to search for other users, right? I think it'd be a good functionality to have honestly, so you can add it if you want to. But this also means that we should allow users to hide their lists/make their profile private if they want, for the sake of privacy.

Btw, just asking, you mentioned something about a blog functionality earlier sometime, is that something like how MAL does? I mean- something like a personal blog where each user can participate with their thoughts and ideas, much like a larger community forum? or was it something else?

I personally think having a blog like forum is a great idea but I'm not sure about how much work/technically hard it is. 

As for the pages, I can design them in any layout, so no worries on that part.

Once again, great work! 
Good day to you

^^

baseplate-admin commented:
> Also, by user-search, do you mean being able to search for other users, right?

Yep. So for this i am doing something like

`http://127.0.0.1:8000/api/v1/user/info`  |   To `GET` / `POST` userinfo
`http://127.0.0.1:8000/api/v1/user/{str:username}` -> `http://127.0.0.1:8000/api/v1/user/baseplate-admin`  |  To get user info.

AKindWorld commented:
Hmm I like the username method more. Remembering usernames are much easier than user-ids, so we can do that.

Just one thing - do you think we can make an integrated approach like: username+tag (similar to Discord), so that more than 1 person can get the same username yet with a different id (by using tags in the username)

I think the approach that way should be:

`http://127.0.0.1:8000/api/v1/user/{str:usernametag}` → `http://127.0.0.1:8000/api/v1/user/Titan0001`

This method is better in one way since, if someone else already has the username Titan, but I want Titan too, so that way we can just get different tags with the same username. (so someone else who wants Titan too gets `http://127.0.0.1:8000/api/v1/user/Titan0002` and so on)

I'm not sure if this is possible though. We can also make people customize their tags if they want to later on (like `http://127.0.0.1:8000/api/v1/user/Titan04` or `http://127.0.0.1:8000/api/v1/user/Titan999` or even remove tags completely from their username - maybe as a bonus feature?)

Is this possible? If not or if it's a lot of work, then feel free to use the existing username system :)

---

> This is not implemented ( yet )

No worries. But I do think it is a somewhat essential feature (atleast later on when the site grows); but take your time!

> Umm we are doing something API-ish ( backend ) . So there wont be any page :D

Ah yup, I meant the frontend designs for the forum and user profiles. I already made a profile design in the legacy files but it needs to updated to the latest design system, so I'll work on it soon.

Great work!

^^

baseplate-admin commented:
> Just one thing - do you think we can make an integrated approach like: username+tag (similar to Discord), so that more than 1 person can get the same username yet with a different id (by using tags in the username)

> I think the approach that way should be:

> http://127.0.0.1:8000/api/v1/user/{str:usernametag} → http://127.0.0.1:8000/api/v1/user/Titan0001

> This method is better in one way since, if someone else already has the username Titan, but I want Titan too, so that way we can just get different tags with the same username. (so someone else who wants Titan too gets http://127.0.0.1:8000/api/v1/user/Titan0002 and so on)

> I'm not sure if this is possible though. We can also make people customize their tags if they want to later on (like http://127.0.0.1:8000/api/v1/user/Titan04 or http://127.0.0.1:8000/api/v1/user/Titan999 or even remove tags completely from their username - maybe as a bonus feature?)

Yep i am looking into it but it will be a tough one

https://github.com/django/django/blob/fd93db97c7228b16a4f92f97ef05b0d72418d952/django/contrib/auth/models.py#L344-L355

See the unique tag? That is the main problem. But lemme see if i can cook up anything new

AKindWorld commented:
> Done

WHOA that was super fast! You're really amazing xD

Now that this is done, I'll design the login & register pages too. 

One question - do we need different pages for login & register? Also, can both login & register be done in modals or do they need separate pages of their own?

Let me know if I should merge this PR.

^^

baseplate-admin commented:
```SQL
 SELECT "__user___user"."id",
       "__user___user"."password",
       "__user___user"."last_login",
       "__user___user"."is_superuser",
       "__user___user"."username",
       "__user___user"."first_name",
       "__user___user"."last_name",
       "__user___user"."email",
       "__user___user"."is_staff",
       "__user___user"."is_active",
       "__user___user"."date_joined",
       "__user___user"."username_discriminator",
       "__user___user"."avatar",
       "__user___user"."video_volume",
       Cast("__user___user"."username_discriminator" AS TEXT) AS "username_discriminator_as_string",
       COALESCE("__user___user"."username", )
              || COALESCE(COALESCE(#, )
              || COALESCE(lpad(cast("__user___user"."username_discriminator" AS text), 4, 0), ), ) AS "username_with_discriminator"
FROM   "__user___user"
WHERE  COALESCE("__user___user"."username", )
              || COALESCE(COALESCE(#, )
              || COALESCE(lpad(cast("__user___user"."username_discriminator" AS text), 4, 0), ), ) IN (baseplate-admin#0001) 
```

This works

---

> One question - do we need different pages for login & register? Also, can both login & register be done in modals or do they need separate pages of their own?

So umm we need to have 2 types of page. One for backend and one for frontend. For frontend we can handle that with modals. For backend we need strict validation and for that a better approach would be a full page registration screen

baseplate-admin commented:
Hey @AKindWorld, 

I am adding artificial intelligence ( more so machine learning ) to our current backend

baseplate-admin commented:
For now lemme merge this :D ( though i did it unintentionally ) 


I will open a new PR in just a sec
